<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [448]:
import pandas as pd
import psycopg2

In [450]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [451]:
# текст запроса
query_3_1 = f'''select COUNT(v.id)
            from vacancies v 
         '''

In [452]:
df = pd.read_sql_query(query_3_1, connection)
df

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [453]:
query = f'''select COUNT(e.id)
            from employers e
         '''

In [454]:
df = pd.read_sql_query(query, connection)
df

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/137330334.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [455]:
query = f'''select COUNT(a.id)
            from areas a
         '''

In [456]:
df = pd.read_sql_query(query, connection)
df

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/137330334.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [457]:
query = f'''select COUNT(i.id)
            from industries i
         '''

In [458]:
df = pd.read_sql_query(query, connection)
df

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/137330334.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    294

***

### Выводы по блоку данных:

#### 1. Общая численность вакансий
Количество вакансий в базе данных составляет 49197. Это указывает на значительный объем данных, что может позволить провести обширный анализ рынка труда, востребованных навыков, географических распределений и других параметров.

#### 2. Количество работодателей
В базе данных содержится информация о 23501 работодателе. Это подчеркивает разнообразие компаний, представленных в базе, и указывает на разнообразие сфер деятельности, в которых эти компании работают. Также это может свидетельствовать о высокой конкуренции на рынке труда.

#### 3. Количество регионов
База данных охватывает 1362 региона. Это свидетельствует о широком географическом охвате данных и позволяет проводить анализ по региональному распределению вакансий и востребованности специалистов в различных частях страны.

#### 4. Количество сфер деятельности
Количество сфер деятельности составляет 294. Это демонстрирует разнообразие отраслей, в которых размещены вакансии, и позволяет оценить востребованность специалистов в различных профессиональных областях.

### Заключение
Этот блок данных предоставляет ценную информацию для анализа рынка труда. Высокое количество вакансий и работодателей указывает на насыщенность базы и потенциальную глубину анализа. Географический и отраслевой охват данных позволяют провести многомерный анализ, выявить тенденции и потребности в различных регионах и отраслях. Эти данные могут быть полезны для специалистов по трудоустройству, аналитиков рынка труда, исследователей и других заинтересованных сторон.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [459]:
query = f'''SELECT a.name area, COUNT(v.id) cnt
            FROM vacancies v
            JOIN areas a ON v.area_id = a.id
            GROUP BY a.name
            ORDER BY 2 DESC
            LIMIT 5
         '''

In [460]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [461]:
query = f'''SELECT COUNT(v.id)
            FROM vacancies v
            WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
         '''

In [462]:
df = pd.read_sql_query(query, connection)
df

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/137330334.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [463]:
query = f'''SELECT ROUND(AVG(v.salary_from)) avg_from, ROUND(AVG(v.salary_to)) avg_to
            FROM vacancies v
            WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
         '''

In [464]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_from    avg_to
0   71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [465]:
query = f'''SELECT v.schedule, v.employment, COUNT(employment)
            FROM vacancies v
            GROUP BY v.schedule, v.employment
            ORDER BY 3 DESC
            OFFSET 1
            LIMIT 1
         '''

In [466]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


schedule        employment  count
0  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [467]:
query = f'''SELECT v.experience, COUNT(v.experience)
            FROM vacancies v
            GROUP BY v.experience
            ORDER BY 2
         '''

In [468]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Детальный анализ вакансий показывает, что крупные города, такие как Москва и Санкт-Петербург, лидируют по количеству предложений работы. Большинство вакансий имеют указанные зарплатные вилки, что позволяет более точно оценивать предложения. Распределение вакансий по типу занятости и требуемому опыту работы показывает, что рынок труда в основном ориентирован на специалистов с опытом от 1 года до 3 лет, а удалённая работа на условиях полной занятости становится все более популярной. Эти данные могут быть полезны для соискателей при выборе работы и для работодателей при формулировании требований к кандидатам.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [469]:
query = f'''(SELECT e.name, COUNT(e.id)
            FROM vacancies v
            JOIN employers e ON v.employer_id = e.id
            GROUP BY e.name
            ORDER BY 2 DESC
            LIMIT 1)
            
            UNION
            
            (SELECT e.name, COUNT(e.id)
            FROM vacancies v
            JOIN employers e ON v.employer_id = e.id
            GROUP BY e.name
            ORDER BY 2 DESC
            OFFSET 4
            LIMIT 1)
            
            ORDER BY 2 DESC
         '''

In [470]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [471]:
query = f'''SELECT a.name region, COUNT(e.id) employers, COUNT(v.id) vacancies
            FROM employers e
            JOIN areas a ON a.id = e.area
            LEFT JOIN vacancies v ON e.id = v.employer_id
            GROUP BY a.name
            HAVING COUNT(v.id) = 0
            ORDER BY 2 DESC
            LIMIT 1
         '''

In [472]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


region  employers  vacancies
0  Серпухов          6          0

не понимаю почему Серпухов не подходит

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [473]:
query = f'''SELECT e.name, COUNT(DISTINCT v.area_id)
            FROM employers e
            JOIN vacancies v ON e.id = v.employer_id
            GROUP BY e.name
            ORDER BY 2 DESC
            LIMIT 1
         '''

In [474]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0  Яндекс    181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [475]:
query = f'''SELECT COUNT(e.id)
            FROM employers e
            LEFT JOIN employers_industries ei ON e.id = ei.employer_id
            WHERE ei.industry_id IS NULL
         '''

In [476]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [477]:
query = f'''SELECT name, COUNT(ei.industry_id)
            FROM employers e
            JOIN employers_industries ei ON e.id = ei.employer_id
            GROUP BY name
            HAVING COUNT(ei.industry_id) = 4
            ORDER BY 1
            OFFSET 2
            LIMIT 1
         '''

In [478]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [479]:
query = f'''SELECT i.name, COUNT(e.id)
            FROM employers e
            JOIN employers_industries ei ON e.id = ei.employer_id
            JOIN industries i ON i.id = ei.industry_id
            GROUP BY i.name
            HAVING i.name = 'Разработка программного обеспечения'
         '''

In [480]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0  Разработка программного обеспечения   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [481]:
from bs4 import BeautifulSoup
import requests

url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')

table = soup.find('table')
links = table.select('tr td:nth-child(2) a')
cities = [link.text for link in links]
cities_tuple = tuple(cities)

In [482]:
query = f'''
(
    SELECT a.name, COUNT(v.id) cnt
    FROM areas a
    LEFT JOIN vacancies v ON a.id = v.area_id
    JOIN employers e ON v.employer_id = e.id
    WHERE e.name = 'Яндекс' AND a.name IN {cities_tuple}
    GROUP BY a.name
)
UNION
(
    SELECT 'Total', COUNT(v.id)
    FROM areas a
    LEFT JOIN vacancies v ON a.id = v.area_id
    JOIN employers e ON v.employer_id = e.id
    WHERE e.name = 'Яндекс' AND a.name IN {cities_tuple}
)
ORDER BY 2;
'''

In [483]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

Анализ работодателей показывает, что крупные компании, такие как Яндекс и Газпром нефть, активно набирают сотрудников. Регионы с наибольшим количеством работодателей, но без вакансий, могут указывать на экономические или административные барьеры. Максимальное количество регионов, в которых публикуются вакансии одного работодателя, свидетельствует о широком охвате бизнеса. Большое количество работодателей в сфере “Разработка программного обеспечения” подчеркивает важность ИТ-индустрии. Анализ вакансий компании “Яндекс” в регионах-миллионниках демонстрирует их значительное присутствие на рынке труда.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [484]:
query = f'''
SELECT COUNT(v.name)
FROM vacancies v
WHERE LOWER(v.name) LIKE '%data%' OR LOWER(v.name) LIKE '%данн%'
'''

In [485]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [486]:
query = f'''
SELECT COUNT(v.name)
FROM vacancies v
WHERE (
    LOWER(v.name) LIKE '%data scientist%' OR
    LOWER(v.name) LIKE '%data science%' OR
    LOWER(v.name) LIKE '%исследователь данных%' OR
    (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%') OR
    LOWER(v.name) LIKE '%machine learning%' OR
    LOWER(v.name) LIKE '%машинн%обучен%'
)
    AND 
    (
        LOWER(v.name) LIKE '%junior%' OR
        v.experience = 'Нет опыта' OR
        v.employment = 'Стажировка'
    )
'''

In [487]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [488]:
query = f'''
SELECT COUNT(v.name)
FROM vacancies v
WHERE (
    LOWER(v.name) LIKE '%data scientist%' OR
    LOWER(v.name) LIKE '%data science%' OR
    LOWER(v.name) LIKE '%исследователь данных%' OR
    (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%') OR
    LOWER(v.name) LIKE '%machine learning%' OR
    LOWER(v.name) LIKE '%машинн%обучен%'
) AND (
        LOWER(v.key_skills) LIKE '%sql%' OR 
        LOWER(v.key_skills) LIKE '%postgres%'
      )
'''

In [489]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [490]:
query = f'''
SELECT COUNT(v.name)
FROM vacancies v
WHERE (
    LOWER(v.name) LIKE '%data scientist%' OR
    LOWER(v.name) LIKE '%data science%' OR
    LOWER(v.name) LIKE '%исследователь данных%' OR
    (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%') OR
    LOWER(v.name) LIKE '%machine learning%' OR
    LOWER(v.name) LIKE '%машинн%обучен%'
) AND (
        LOWER(v.key_skills) LIKE '%python%'
      )
'''

In [491]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [492]:
query = f'''
SELECT ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, '\t', '')) + 1), 2) AS avg_skills
FROM vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%' OR
    LOWER(name) LIKE '%data science%' OR
    LOWER(name) LIKE '%исследователь данных%' OR
    (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR
    LOWER(name) LIKE '%machine learning%' OR
    LOWER(name) LIKE '%машинн%обучен%'
) AND key_skills IS NOT NULL
AND key_skills != ''
'''

In [493]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_skills
0        6.55

понятия не имею почему ответ неверный. Я пробовал и chr(9) и \t результат один и тот же



6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [494]:
query = f'''
SELECT experience, ROUND(AVG(
    CASE
        WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
        WHEN salary_from IS NOT NULL THEN salary_from
        WHEN salary_to IS NOT NULL THEN salary_to
    END), 0) AS avg_salary
FROM vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%' OR
    LOWER(name) LIKE '%data science%' OR
    LOWER(name) LIKE '%исследователь данных%' OR
    (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR
    LOWER(name) LIKE '%machine learning%' OR
    LOWER(name) LIKE '%машинн%обучен%'
)
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY experience
'''

In [495]:
df = pd.read_sql_query(query, connection)
display(df)

/var/folders/dl/7gg59dmd491c8_bgddzqs7p40000gq/T/ipykernel_19025/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience  avg_salary
0         Более 6 лет    157933.0
1       От 3 до 6 лет    256454.0
2           Нет опыта     74643.0
3  От 1 года до 3 лет    144233.0

все проверил 100 раз ответ неверный. не понимаю почему

***

Предметный анализ показывает, что вакансии, связанные с данными, составляют значительную часть рынка труда, а навыки работы с SQL, Postgres и Python высоко востребованы. Среднее количество ключевых навыков, указанных в вакансиях для дата-сайентистов, указывает на необходимость владения широким спектром компетенций. Высокий уровень зарплат для специалистов с опытом работы от 3 до 6 лет демонстрирует привлекательность профессии дата-сайентиста на рынке труда. Эти данные могут быть полезны как для начинающих, так и для опытных специалистов при планировании своей карьеры и повышения квалификации.

# Общий вывод по проекту

In [496]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

In [497]:
connection.close()